### PR TITLE
[llvm][Support] Add UNIX socket support

### PR DESCRIPTION
**Pull request by criis**
This adds support for UNIX socket communication to work similarly to raw_stream.

### CODE REVIEW

**Review by Bigcheese — CHANGES_REQUESTED** [review]
Thanks for the patch. As a high level comment you should include a bit more information in the commit message.

Also the llvm/lib/Support/CMakeLists.txt file needs to change line 44 to:
`set(system_libs ${system_libs} psapi shell32 ole32 uuid advapi32 Ws2_32)`

This links against the winsock library.

File: llvm/lib/Support/raw_ostream.cpp [hunk 1]
```diff
@@ -942,6 +959,146 @@ bool raw_fd_stream::classof(const raw_ostream *OS) {
...
+  if (WSAStartup(MAKEWORD(2, 2), &WsaData) != 0) {
+    llvm::report_fatal_error("WSAStartup failed");
+  }
+}
+
+WSABalancer::~WSABalancer() { WSACleanup(); }
+
+#endif // _WIN32
+
+std::error_code getLastSocketErrorCode() {
```
Comment (Bigcheese) [inline llvm/lib/Support/raw_ostream.cpp:978]: This should be `static` as it's only used in this file.

File: llvm/lib/Support/raw_ostream.cpp [hunk 2]
```diff
@@ -942,6 +959,146 @@ bool raw_fd_stream::classof(const raw_ostream *OS) {
...
+
+ListeningSocket::~ListeningSocket() {
+  if (FD == -1)
+    return;
+  ::close(FD);
+  unlink(SocketPath.c_str());
+}
+
+// Expected?
+Expected<int> GetSocketFD(StringRef SocketPath) {
```
Comment (Bigcheese) [inline llvm/lib/Support/raw_ostream.cpp:1057]: This should be `static` as it's only used in this file.

**Review by Bigcheese — COMMENTED** [review]
I think it looks good with these last few changes.

**Review by Bigcheese — APPROVED** [review]

File: llvm/include/llvm/Support/raw_ostream.h [hunk 3]
```diff
@@ -16,6 +16,7 @@
 #include "llvm/ADT/SmallVector.h"
 #include "llvm/ADT/StringRef.h"
 #include "llvm/Support/DataTypes.h"
+#include "llvm/Support/Threading.h"
```
Comment (ributzka) [inline llvm/include/llvm/Support/raw_ostream.h:19]: Is this include required in the header?

**Comment by yozhu** [comment]
Hit compilation errors when building LLVM for Windows:

llvm/lib/Support/raw_ostream.cpp(1018,22): error: variable has incomplete type 'struct sockaddr_un'
[CONTEXT]  1018 |   struct sockaddr_un Addr;
[CONTEXT]       |                      ^

llvm/lib/Support/raw_ostream.cpp(1077,22): error: variable has incomplete type 'struct sockaddr_un'
[CONTEXT]  1077 |   struct sockaddr_un Addr;
[CONTEXT]       |                      ^

**Comment by Bigcheese** [comment]
> Hit compilation errors when building LLVM for Windows:
> 
> llvm/lib/Support/raw_ostream.cpp(1018,22): error: variable has incomplete type 'struct sockaddr_un' [CONTEXT] 1018 | struct sockaddr_un Addr; [CONTEXT] | ^
> 
> llvm/lib/Support/raw_ostream.cpp(1077,22): error: variable has incomplete type 'struct sockaddr_un' [CONTEXT] 1077 | struct sockaddr_un Addr; [CONTEXT] | ^

What version of MSVC do you have? This should be supported by the minimum supported version.

**Comment by shiltian** [comment]
Am I the only one getting the following error?
```
/home/ac.shilei.tian/Documents/vscode/llvm-project/llvm/lib/Support/raw_ostream.cpp: In static member function ‘static llvm::Expected<llvm::ListeningSocket> llvm::ListeningSocket::createUnix(llvm::StringRef, int)’:
/home/ac.shilei.tian/Documents/vscode/llvm-project/llvm/lib/Support/raw_ostream.cpp:1040:10: error: could not convert ‘ListenSocket’ from ‘llvm::ListeningSocket’ to ‘llvm::Expected<llvm::ListeningSocket>’
   return ListenSocket;
          ^~~~~~~~~~~~
```

**Comment by Bigcheese** [comment]
> Am I the only one getting the following error?
> 
> ```
> /home/ac.shilei.tian/Documents/vscode/llvm-project/llvm/lib/Support/raw_ostream.cpp: In static member function ‘static llvm::Expected<llvm::ListeningSocket> llvm::ListeningSocket::createUnix(llvm::StringRef, int)’:
> /home/ac.shilei.tian/Documents/vscode/llvm-project/llvm/lib/Support/raw_ostream.cpp:1040:10: error: could not convert ‘ListenSocket’ from ‘llvm::ListeningSocket’ to ‘llvm::Expected<llvm::ListeningSocket>’
>    return ListenSocket;
>           ^~~~~~~~~~~~
> ```
> 
> Compiler information:
> 
> ```
> ➜  /usr/bin/c++ --version
> c++ (SUSE Linux) 7.5.0
> Copyright (C) 2017 Free Software Foundation, Inc.
> This is free software; see the source for copying conditions.  There is NO
> warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
> ```

Assuming that's gcc 7.5, looks like it doesn't do all of [P0135](https://wg21.link/P0135). That should be easy to change the code to support.

**Comment by shiltian** [comment]
> > Am I the only one getting the following error?
> > ```
> > /home/ac.shilei.tian/Documents/vscode/llvm-project/llvm/lib/Support/raw_ostream.cpp: In static member function ‘static llvm::Expected<llvm::ListeningSocket> llvm::ListeningSocket::createUnix(llvm::StringRef, int)’:
> > /home/ac.shilei.tian/Documents/vscode/llvm-project/llvm/lib/Support/raw_ostream.cpp:1040:10: error: could not convert ‘ListenSocket’ from ‘llvm::ListeningSocket’ to ‘llvm::Expected<llvm::ListeningSocket>’
> >    return ListenSocket;
> >           ^~~~~~~~~~~~
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Compiler information:
> > ```
> > ➜  /usr/bin/c++ --version
> > c++ (SUSE Linux) 7.5.0
> > Copyright (C) 2017 Free Software Foundation, Inc.
> > This is free software; see the source for copying conditions.  There is NO
> > warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
> > ```
> 
> Assuming that's gcc 7.5, looks like it doesn't do all of [P0135](https://wg21.link/P0135). That should be easy to change the code to support.

Please do as it is blocking my build and GCC 7.5.0 is definitely supported by LLVM right now.

**Comment by yozhu** [comment]
> > Hit compilation errors when building LLVM for Windows:
> > llvm/lib/Support/raw_ostream.cpp(1018,22): error: variable has incomplete type 'struct sockaddr_un' [CONTEXT] 1018 | struct sockaddr_un Addr; [CONTEXT] | ^
> > llvm/lib/Support/raw_ostream.cpp(1077,22): error: variable has incomplete type 'struct sockaddr_un' [CONTEXT] 1077 | struct sockaddr_un Addr; [CONTEXT] | ^
> 
> What version of MSVC do you have? This should be supported by the minimum supported version.

We are currently using WinSDK 10.0.16299.0. What version would you recommend we move to?

**Comment by Bigcheese** [comment]
> We are currently using WinSDK 10.0.16299.0. What version would you recommend we move to?

I think 10.0.17134.0 which is included in VS2017 ver.15.7.

However, the minimum currently supported VS is VS2019 as of https://reviews.llvm.org/D114639, which means at least Windows SDK 10.0.18362.0.

**Comment by nico** [comment]
From what I understand, this makes binaries depending on Support, such as clang, now depend on winsock. That seems like an undesirable dependency in a compiler. Maybe this could live in its own library, instead of in Support? Then only tools that need unix sockets could depend on it.

**Comment by Bigcheese** [comment]
> From what I understand, this makes binaries depending on Support, such as clang, now depend on winsock. That seems like an undesirable dependency in a compiler. Maybe this could live in its own library, instead of in Support? Then only tools that need unix sockets could depend on it.

I considered that a bit, but we already depend on stuff like shell32 and COM, and just linking to winsock doesn't really do anything, you have to actually call `WSAStartup` before it sets up anything.

If anyone actually encounters an issue with linking against winsock then it'd be fine with moving it, but I don't really see a point in doing that right now. Also note that clang itself will depend on this once https://github.com/llvm/llvm-project/pull/67562 lands. It was waiting on Windows support which is what this patch was for.

**Comment by nikic** [comment]
Can you please move this out of raw_ostream.h into a separate header? This is a core header, and adding extra dependencies to it has visible effect on time to compile clang.

**Comment by nico** [comment]
> I considered that a bit, but we already depend on stuff like shell32 and COM, and just linking to winsock doesn't really do anything, you have to actually call WSAStartup before it sets up anything.

It does show up in `dumpbin /dependents` output, which is super weird for a compiler. And now, if you have a compile farm, an attacker might be able to send you source code that can just call dormant winsock functions to do network access. (Granted, previously they could LoadLibrary() it, and you should have a sandbox, but it's still bad for a compiler to have a dep on winsock.) And for the other deps, we at least delayload them.

https://github.com/llvm/llvm-project/pull/67562 also looks like something that should be in a separate binary, not in clang itself.

**Comment by Bigcheese** [comment]
> It does show up in `dumpbin /dependents` output, which is super weird for a compiler. And now, if you have a compile farm, an attacker might be able to send you source code that can just call dormant winsock functions to do network access. (Granted, previously they could LoadLibrary() it, and you should have a sandbox, but it's still bad for a compiler to have a dep on winsock.) And for the other deps, we at least delayload them.

I think it's fine to delay load it, I'll try out adding that.

As for the security concerns, it's already true that you have to lock down clang (or almost any compiler) if you're accepting arbitrary code. On most POSIX systems you already have access to the socket functions via libc anyway, and as you say on Windows you can just dynamically load them. Linking against winsock doesn't actually change anything here.

Note that GCC also uses sockets now for modules via `-fmodule-mapper`, including support for TCP.

**Pull request merged**